RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc

```diff
@@ -78,298 +94,301 @@ void TSGForOI::produce(edm::StreamID sid, edm::Event& iEvent, const edm::EventSe
+  for (unsigned int l2TrackColIndex(0); l2TrackColIndex!=l2TrackCol->size(); ++l2TrackColIndex) {
+
     const reco::TrackRef l2(l2TrackCol, l2TrackColIndex);
-    std::unique_ptr<std::vector<TrajectorySeed> > out(new std::vector<TrajectorySeed>());
+
+    // Container of Seeds
+    std::vector<TrajectorySeed> out;
     LogTrace("TSGForOI") << "TSGForOI::produce: L2 muon pT, eta, phi --> " << l2->pt() << " , " << l2->eta() << " , " << l2->phi() << endl;
-    
+ 
     FreeTrajectoryState fts = trajectoryStateTransform::initialFreeState(*l2, magfieldH.product());
+
     dummyPlane->move(fts.position() - dummyPlane->position());
     TrajectoryStateOnSurface tsosAtIP = TrajectoryStateOnSurface(fts, *dummyPlane);
     LogTrace("TSGForOI") << "TSGForOI::produce: Created TSOSatIP: " << tsosAtIP << std::endl;
     
-    // get the TSOS on the innermost layer of the L2. 
+    // Get the TSOS on the innermost layer of the L2
     TrajectoryStateOnSurface tsosAtMuonSystem = trajectoryStateTransform::innerStateOnSurface(*l2, *geometryH, magfieldH.product());
     LogTrace("TSGForOI") << "TSGForOI::produce: Created TSOSatMuonSystem: " << tsosAtMuonSystem <<endl;
     
     LogTrace("TSGForOI") << "TSGForOI::produce: Check the error of the L2 parameter and use hit seeds if big errors" << endl;
+
     StateOnTrackerBound fromInside(propagatorAlong.get());
     TrajectoryStateOnSurface outerTkStateInside = fromInside(fts);
-      
-    StateOnTrackerBound fromOutside(&*SmartOpposite);
+
+    StateOnTrackerBound fromOutside(&*SHPOpposite);
     TrajectoryStateOnSurface outerTkStateOutside = fromOutside(tsosAtMuonSystem);
 
-      // for now only checking if the two positions (using updated and not-updated) agree withing certain extent, 
-      // will probably have to design something fancier for the future. 
-    auto dist=0.0;
+    // Check if the two positions (using updated and not-updated TSOS) agree withing certain extent.
+    // If both TSOSs agree, use only the one at vertex, as it uses more information. If they do not agree, search for seeds based on both.
     bool useBoth = false;
-    if (outerTkStateInside.isValid() && outerTkStateOutside.isValid()){
-	dist = match_Chi2(outerTkStateInside,outerTkStateOutside);
-    }
-    if (dist>tsosDiff_){
-	useBoth = true; 
+    if (outerTkStateInside.isValid() && outerTkStateOutside.isValid()) {
+      //auto dist1 = match_Chi2(outerTkStateInside,outerTkStateOutside);
+      //auto dist2 = deltaR(outerTkStateInside.globalMomentum(),outerTkStateOutside.globalMomentum());
+      //if ((dist1 > tsosDiff1_ || dist2 > tsosDiff2_) && l2->numberOfValidHits() < 20) useBoth = true;
+      if (l2->numberOfValidHits() < 20) useBoth = true;
+      if (l2->numberOfValidHits() < 30 && (std::abs(l2->eta()) > 2.1)) useBoth = true;
+      //if (std::abs(l2->eta())<1.0)useBoth=false;
+      if (std::abs(l2->eta()) > 0.2 && std::abs(l2->eta()) < 0.3) useBoth = true;
```

Comment:
it seems like some of these magic numbers can be made configurable.
I can imagine that the same requirements will not apply in Run-3 or phase-2

Reply:
Okay .. could be made configurable